### ISSUE TITLE
Leap Year effect

### ISSUE THREAD

**Issue by froehlich1** [jdemetra/rjdemetra#67]
Hi,

I run an X13 adjustment with the following specification:
x13_spec(spec = "RSA3", transform.function = "None", 
                 usrdef.varEnabled = TRUE, 
                 usrdef.var = tdProd, usrdef.varType = "Calendar", 
                 tradingdays.option = "UserDefined",
                 tradingdays.autoadjust = FALSE,
                 tradingdays.leapyear = "LeapYear"

However, I am not sure, if the leapYear adjustment is actually performed. At least I cannot see any leapYear regressor in the regression output. 
Can I access the leap year regresser somehow, or did I forget some options?

Best,

**Comment by AQLT** [comment]
Hi,
With the option `tradingdays.option = "UserDefined"` you have to add the leap year regressor in your user defined variables: so the leap year adjustment is not done with your specification. Normally, if we have correctly implemented the functions, you should, printing your specification, see that the one specific to the leap year are not changed. In any case I will try to add a warning message to prevent this error.

**Comment by AQLT** [comment]
A warning is now added (#70). This is also in the last release version in CRAN.

**Comment by merangelik** [comment]
I am running into problems including the leap year:
https://github.com/statistikat/persephone/issues/21

Am I missing some additional parameters that need to be set?

**Comment by froehlich1** [comment]
Hi, der leapyearregressor muss bereits bei den trading days dabei sein.
Wenn du genTd verwendest ist das inkludiert
Lg

Angelika Meraner <notifications@github.com> schrieb am Di., 7. Apr. 2020,
15:37:

**Comment by ClementFeltin** [comment]
> 
> 
> Hi,
> With the option `tradingdays.option = "UserDefined"` you have to add the leap year regressor in your user defined variables: so the leap year adjustment is not done with your specification. Normally, if we have correctly implemented the functions, you should, printing your specification, see that the one specific to the leap year are not changed. In any case I will try to add a warning message to prevent this error.

Hello,
Could you provide with an example ? I am not sure what a leap year or Easter effect regressor looks like.
Thanks :)

**Comment by AQLT** [comment]
Hi,
For a theoretical description of the calendar regressors you can for example look at the handbook on Seasonal adjustment ([https://ec.europa.eu/eurostat/fr/web/products-manuals-and-guidelines/-/KS-GQ-18-001](url)), chapter 5.
If you want to see how the regressors looks like, the most simple way is to open JDemetra+, to open a series and go to the Pre-processing > Regressors. The leap year regressor is simple and can be easily implemented in R: it is equal to 0,75 in February during leap years, -0.25 in February during non leap years and 0 otherwise. The gradual easter regressor is a bit more complicated but when you use the option `tradingdays.option = "UserDefined"` you don't need to add it (the `easter.xxx` options remains available).
For the leap year regressor, indeed you should use `varType = "Calendar"`. For example:
``` r
LY_regressor <- function(start_year = 1970, end_year = 2020, frequency = 12){
  reg <- ts(0,start = c(start_year,1),
            end = c(end_year,12), frequency = 12)
  all_feb <- which(cycle(reg) == 2)
  if (length(all_feb) > 0){
    for (feb in all_feb){
      year <- trunc(time(reg))[feb]
      reg[feb] <- ifelse((year%%4 == 0 & year%%100 != 0) | (year%%400 == 0),
                         0.75, -0.25)
    }
  }
  return(aggregate(reg,nfrequency = frequency))
}
ly_r <- LY_regressor(frequency = 12)
ly_r
#>        Jan   Feb   Mar   Apr   May   Jun   Jul   Aug   Sep   Oct   Nov   Dec
#> 1970  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1971  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1972  0.00  0.75  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1973  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1974  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1975  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1976  0.00  0.75  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1977  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1978  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1979  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1980  0.00  0.75  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1981  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1982  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1983  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1984  0.00  0.75  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1985  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1986  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1987  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1988  0.00  0.75  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1989  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1990  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1991  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1992  0.00  0.75  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1993  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1994  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1995  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1996  0.00  0.75  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1997  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1998  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 1999  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 2000  0.00  0.75  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 2001  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 2002  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 2003  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 2004  0.00  0.75  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 2005  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 2006  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 2007  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 2008  0.00  0.75  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 2009  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 2010  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 2011  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 2012  0.00  0.75  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 2013  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 2014  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 2015  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 2016  0.00  0.75  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 2017  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 2018  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 2019  0.00 -0.25  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
#> 2020  0.00  0.75  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00  0.00
library(RJDemetra)
myseries <- ipi_c_eu[, "FR"]
# unrealistic calendar regressors
var1 <- ts(rnorm(length(myseries))*10, start = start(myseries), frequency = 12)
var2 <- ts(rnorm(length(myseries))*100, start = start(myseries), frequency = 12)
var <- ts.union(var1, var2, ly_r)
myspec1 <- x13_spec(spec = "RSA5c", tradingdays.option = "UserDefined",
                    tradingdays.test = "None",
                    usrdef.varEnabled = TRUE,
                    usrdef.var = var,
                    usrdef.varType = c("Calendar"))
#> Warning: With tradingdays.option = "UserDefined", the parameters tradingdays.autoadjust, tradingdays.leapyear and tradingdays.stocktd are ignored.
myreg1 <- x13(myseries, myspec1)
myreg1$regarima
#> y = regression model + arima (3, 1, 1, 0, 1, 1)
#> Log-transformation: yes
#> Coefficients:
#>           Estimate Std. Error
#> Phi(1)     0.34499      0.363
#> Phi(2)     0.07678      0.364
#> Phi(3)    -0.42332      0.263
#> Theta(1)  -0.67115      0.350
#> BTheta(1) -0.72715      0.040
#> 
#>                Estimate Std. Error
#> var1          1.418e-04      0.000
#> var2          1.917e-05      0.000
#> ly_r          1.672e-02      0.008
#> Easter [8]   -2.379e-02      0.006
#> LS (11-2008) -8.980e-02      0.017
#> AO (5-2011)   9.043e-02      0.021
#> LS (1-2009)  -7.640e-02      0.017
#> 
#> 
#> Residual standard error: 0.02667 on 334 degrees of freedom
#> Log likelihood = 759.8, aic =  1712 aicc =  1713, bic(corrected for length) = -7.046
```

<sup>Created on 2020-11-12 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>
